File: assignment1/src/Dijkstra.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.PriorityQueue;

public class Dijkstra {
    private final Graph graph;
    private final int numNodes;
    private final double[] dist;
    private final boolean[] visited;
    private final PriorityQueue<NodeDistance> pq;

    public Dijkstra(Graph graph) {
        this.graph = graph;
        this.numNodes = graph.nodes.length;
        this.dist = new double[numNodes];
        this.visited = new boolean[numNodes];
        this.pq = new PriorityQueue<>(Comparator.comparingDouble(node -> node.dist));
    }

    public double shortestPath(int source, int target) {
        Arrays.fill(dist, Double.POSITIVE_INFINITY);
        Arrays.fill(visited, false);
        pq.clear();

        dist[source] = 0;
        pq.add(new NodeDistance(source, 0));

        while (!pq.isEmpty()) {
            NodeDistance current = pq.poll();
            int u = current.node;

            if (visited[u]) continue;
            visited[u] = true;

            if (u == target) return dist[u];

            for (int i = graph.outOffsets[u]; i < graph.outOffsets[u + 1]; i++) {
                Edge edge = graph.outEdges[i];
                int v = edge.to;
                double alt = dist[u] + edge.weight;

                if (!visited[v] && alt < dist[v]) {
                    dist[v] = alt;
                    pq.add(new NodeDistance(v, alt));
                }
            }
        }

        return Double.POSITIVE_INFINITY; // unreachable
    }

    private static class NodeDistance {
        int node;
        double dist;

        NodeDistance(int node, double dist) {
            this.node = node;
            this.dist = dist;
        }
    }
}

File: assignment1/src/dijkstra.rs
use std::cmp::Reverse;
use std::collections::BinaryHeap;
use crate::graph::Graph;


pub struct Dijkstra<'a> {
    source: Option<u32>,
    graph: &'a Graph,
    dist: Vec<u32>,
    heap: BinaryHeap<Reverse<(u32, u32)>>,
}


impl<'a> Dijkstra<'a> {
    pub fn new(graph: &'a Graph) -> Self {
        Dijkstra {
            source: None,
            graph,
            dist: vec![u32::MAX; graph.num_nodes()],
            heap: BinaryHeap::new(),
        }
    }

    pub fn query(&mut self, source: u32, target: u32) -> Option<u32> {
        if self.source != Some(source) {
            self.dist.fill(u32::MAX);
            self.heap.clear();
            self.dist[source as usize] = 0;
            self.heap.push(Reverse((0, source)));
            self.source = Some(source);
        }

        if self.dist[target as usize] != u32::MAX {
            return Some(self.dist[target as usize]);
        }

        while let Some(Reverse((cost, u))) = self.heap.pop() {
            if u == target { return Some(cost); }

            if cost > self.dist[u as usize] { continue; }

            for edge in self.graph.out_edges(u) {
                let next_cost = cost + edge.weight;
                if next_cost < self.dist[edge.target as usize] {
                    self.dist[edge.target as usize] = next_cost;
                    self.heap.push(Reverse((next_cost, edge.target)));
                }
            }
        }

        None
    }
}
